staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -671,19 +744,41 @@ func (c *Cacher) processEvent(event *watchCacheEvent) {
 }
 
 func (c *Cacher) dispatchEvents() {
+	startID := bucketID(c.clock.Now())
 	for {
 		select {
 		case event, ok := <-c.incoming:
 			if !ok {
 				return
 			}
-			c.dispatchEvent(&event)
+			if event.Type == watch.Bookmark {
+				currentID := bucketID(c.clock.Now())
+				c.dispatchBookmarkEvent(&event, startID, currentID)
+				startID = currentID + 1
+			} else {
+				c.dispatchEvent(&event)
+			}
 		case <-c.stopCh:
 			return
 		}
 	}
 }
 
+func (c *Cacher) dispatchBookmarkEvent(event *watchCacheEvent, startID, currentID int64) {
+	c.Lock()
```

Comment:
This may potentially take non-trivial amount of time - is it possible to reduce time for which we take this lock?

Reply:
> non-trivial amount of time

Do you mean gc pause, we will not block in the lock scope?

Reply:
To be clear - I don't think fixing that is a blocker, I just throw a thought about that.

GC is one option, second option is large number of watchers, another is allocation coming from "pushToNextTimeBucket".